textutils_internal.h

```diff
@@ -171,6 +171,125 @@ static inline int isupper_c(char c) { return isupper((unsigned char) c); }
 static inline char tolower_c(char c) { return tolower((unsigned char) c); }
 static inline char toupper_c(char c) { return toupper((unsigned char) c); }
 
+// Faster replacements for strtol, for use when parsing lots of numbers.
+// Note that these only handle base 10 and do not skip leading whitespace
+
+/// Convert a string to a signed integer, with overflow detection
+/** @param[in]  in     Input string
+    @param[out] end    Returned end pointer
+    @param[in]  bits   Bits available for the converted value
+    @param[out] failed Location of overflow flag
+    @return String value converted to an int64_t
+
+Converts a signed decimal string to an int64_t.  The string should
+consist of an optional '+' or '-' sign followed by one or more of
+the digits 0 to 9.  The output value will be limited to fit in the
+given number of bits (including the sign bit).  If the value is too big,
+the largest possible value will be returned and *failed will be set to 1.
+
+The address of the first character following the converted number will
+be stored in *end.
+
+Both end and failed must be non-NULL.
+ */
+static inline int64_t hts_str2int(const char *in, char **end, int bits,
+                                    int *failed) {
+    uint64_t n = 0, limit = (1ULL << (bits - 1)) - 1;
+    uint32_t fast = (bits - 1) * 1000 / 3322 + 1; // log(10)/log(2) ~= 3.322
+    const unsigned char *v = (const unsigned char *) in;
+    const unsigned int ascii_zero = '0'; // Prevents conversion to signed
+    unsigned char d;
+    int neg = 1;
+
+    switch(*v) {
+    case '-':
+        neg=-1;
+        limit++; /* fall through */
+    case '+':
+        v++;
+        break;
+    default:
+        break;
+    }
+
+    while (--fast && *v>='0' && *v<='9')
+        n = n*10 + *v++ - ascii_zero;
+
+    if (!fast) {
+        uint64_t limit_d_10 = limit / 10;
+        uint64_t limit_m_10 = limit - 10 * limit_d_10;
+         while ((d = *v - ascii_zero) < 10) {
+            if (n < limit_d_10 || (n == limit_d_10 && d <= limit_m_10)) {
+                n = n*10 + d;
+                v++;
+            } else {
+                do { v++; } while (*v - ascii_zero < 10);
+                n = limit;
+                *failed = 1;
+                break;
+            }
+        }
+    }
+
+    *end = (char *)v;
```

Comment:
`strtoll` allows the end pointer to be null, in which case it is not used. We should probably implement a similar behaviour here.

Reply:
I didn't bother to do that as every use of this function has a non-NULL `end`.  Mostly it's passing back a pointer to the location after the number so whatever parser called it can continue.  Are there any places where we might want to use this that do pass `NULL` here?

Reply:
If we don't see any use beyond the current ones, considering also that is not a public method, I suppose there is no need. I was thinking of a scenario where `hts_str2int` was called inside a `kstrtok` loop.